src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -241,18 +282,21 @@ private bool MatchRouteValue(ActionDescriptor action, MvcEndpointInfo endpointIn
                     return true;
                 }
 
-                var matchingParameter = endpointInfo.ParsedTemplate.Parameters.SingleOrDefault(p => string.Equals(p.Name, routeKey, StringComparison.OrdinalIgnoreCase));
+                var matchingParameter = endpointInfo.ParsedPattern.Parameters.SingleOrDefault(p => string.Equals(p.Name, routeKey, StringComparison.OrdinalIgnoreCase));
                 if (matchingParameter != null)
                 {
                     // Check that the value matches against constraints on that parameter
                     // e.g. For {controller:regex((Home|Login))} the controller value must match the regex
-                    //
-                    // REVIEW: This is really ugly
-                    if (endpointInfo.Constraints.TryGetValue(routeKey, out var constraint)
-                        && !constraint.Match(new DefaultHttpContext() { RequestServices = _serviceProvider }, NullRouter.Instance, routeKey, new RouteValueDictionary(action.RouteValues), RouteDirection.IncomingRequest))
+                    if (endpointInfo.Constraints.TryGetValue(routeKey, out var constraints))
                     {
-                        // Did not match constraint
-                        return false;
+                        foreach (var constraint in constraints)
+                        {
+                            if (!constraint.Match(_httpContextInstance, NullRouter.Instance, routeKey, new RouteValueDictionary(action.RouteValues), RouteDirection.IncomingRequest))
```

Comment:
What would you think about passing in `null` here for the `httpContext`. In order to be used with endpoint routing a constraint must be able to accept a null httpcontext (link generation).

Reply:
Is the issue here the `exists` constraint and services?

Reply:
Every constraint implementation I have seen throws an exception if httpContext is null.

Reply:
e.g. https://github.com/aspnet/Routing/blob/release/2.2/src/Microsoft.AspNetCore.Routing/Constraints/IntRouteConstraint.cs#L23

Reply:
OK, we need to fix those then. I'll log an issue.

Reply:
+KnownRouteValueConstraint uses HttpContext to access `HttpContext.RequestServices`.

Reply:
I only have myself to blame for this. That's what I was hoping we could do to address it. 

I logged an issue to address the issue of httpcontext globally, so I guess it's up to you if you want to try and address KnownRVC in this PR.

Reply:
This PR is big enough. It can be part of that issue.